docs/tools/terraform/howto/upgrade-provider-v3-v4.rst

```diff
@@ -0,0 +1,92 @@
+Upgrade the Aiven Terraform Provider from v3 to v4
+===================================================
+
+The Aiven Terraform Provider version 4.0.0 was released in February 2023. If you previously upgraded the Terraform provider from version 2 to version 3, then no other actions are needed. Otherwise, this article has information on how to upgrade to version 4. 
+
+Major changes in v4
+''''''''''''''''''''
+
+Aiven Terraform Provider has a `detailed changelog <https://github.com/aiven/terraform-provider-aiven/blob/main/CHANGELOG.md>`_ but the main changes in v4 are:
+
+- schema fields use strict types instead of string
+- support for strict types in diff functions
+
+These deprecated resources have also been removed:
+
+- ``aiven_database``
+- ``aiven_service_user``
+- ``aiven_vpc_peering_connection``
+- ``aiven_flink_table``
+- ``aiven_flink_job``
+
+Upgrade Aiven Terraform provider
+''''''''''''''''''''''''''''''''
+
+You update the Aiven Terraform Provider by editing the providers block of your script. If the version was already set to ``>= 3.0.0`` then the upgrade is automatic.
+
+.. code:: terraform
+    
+    terraform {
+      required_providers {
+        aiven = {
+          source  = "aiven/aiven"
+          version = ">= 4.0.0"
+        }
+      }
+    }
+
+.. tip::
+    You might need to run ``terraform init -upgrade`` for the provider version upgrade to take place.
+    
+Update resource syntax
+''''''''''''''''''''''''
+
+The deprecated fields listed in the major changes were removed. The following example shows how to migrate these fields safely without destroying existing resources.
+
+.. tip::
+    Backup your Terraform state file ``terraform.tfstate`` (if available), just in case of potential rollback.
+
+
+In this example, the ``aiven_database`` field is updated to the service-specific ``aiven_pg_database`` field for an Aiven for PostgreSQL® service. A list of all resources is available in the `Aiven Operator for Terraform documentation <https://registry.terraform.io/providers/aiven/aiven/latest/docs/resources/>`_.
+
+1. Update ``aiven_database`` references to ``aiven_pg_database`` as in this example file:
+
+.. code::
+
+    - resource "aiven_database" "mydatabase" {
+        project       = aiven_project.myproject.project
+        service_name  = aiven_pg.mypg.service_name
+        database_name = "<DATABASE_NAME>"
+    }
+
+
+    + resource "aiven_pg_database" "mydatabase" {
+        project       = aiven_project.myproject.project
+        service_name  = aiven_pg.mypg.service_name
+        database_name = "<DATABASE_NAME>"
+    }
+
+2. View a list of all resources in the state file::
```

Comment:
is it okay that we have double semicolon at the end here? also, if it is, please check step 1 as it doesn't have it that way

Reply:
I'm pretty sure step 1 doesn't have it because it has the code block explicitly as `.. code::` (I just reused the content from the other article). I can change all of these to be consistent, though.

Reply:
ahh I see, well you're welcome if you want to, but it's good as it is already

Reply:
Ok, cool. Thank you!